File: apps/api/app/api/v1/routes/claims.py
from typing import Optional

from fastapi import APIRouter, Depends, File, Form, HTTPException, Query, UploadFile

from app.core.config import Settings, get_settings
from app.domain.claims.adjudication_service import AdjudicationService
from app.domain.claims.confidence_service import ConfidenceService
from app.domain.claims.demo_data import get_demo_outpatient_claim
from app.domain.claims.document_intake_service import ClaimDocumentIntakeError, ClaimDocumentIntakeService
from app.domain.claims.intake_service import IntakeService
from app.domain.claims.models import (
    ClaimDocumentIntakeResponse,
    ClaimProcessingResponse,
    ClaimRecordSummary,
    ClaimReviewRequest,
    ClaimSubmission,
    X12BatchClaimResult,
    X12BatchUploadResponse,
)
from app.domain.claims.payer_verification_service import PayerVerificationService
from app.domain.claims.policy_retrieval_service import PolicyRetrievalService
from app.domain.claims.repository import ClaimsRepository, get_claims_repository
from app.domain.claims.validation_service import ValidationService
from app.domain.claims.x12_parser import X12ParseError, X12ProfessionalClaimParser

router = APIRouter(prefix="/claims", tags=["claims"])

intake_service = IntakeService()
validation_service = ValidationService()
policy_service = PolicyRetrievalService()
adjudication_service = AdjudicationService()
confidence_service = ConfidenceService()
x12_parser = X12ProfessionalClaimParser()
document_intake_service = ClaimDocumentIntakeService()
payer_verification_service = PayerVerificationService()


def _run_claim_processing(
    claim: ClaimSubmission,
    settings: Settings,
    repository: ClaimsRepository,
) -> ClaimProcessingResponse:
    normalized_claim = intake_service.normalize_claim(claim)
    validation = validation_service.validate(normalized_claim)
    matched_policies = policy_service.retrieve(normalized_claim)
    provider_context = None
    utilization_context = adjudication_service.build_utilization_context(normalized_claim)
    providers_repository = getattr(repository, "providers_repository", None)
    build_provider_context = getattr(repository, "_build_provider_context", None)
    if providers_repository is not None and callable(build_provider_context):
        tenant = providers_repository.ensure_tenant(payer_name=normalized_claim.payer_name)
        provider_key = normalized_claim.rendering_provider_id or normalized_claim.provider_id
        provider_name = normalized_claim.rendering_provider_name or normalized_claim.provider_name
        provider = providers_repository.ensure_provider_for_tenant(
            tenant_id=tenant.id or "",
            provider_id=provider_key,
            provider_name=provider_name,
            specialty=None,
        )
        provider_context = build_provider_context(claim=normalized_claim, provider=provider)
    payer_verification_context = payer_verification_service.build(
        normalized_claim,
        provider_context=provider_context,
    )
    decision = adjudication_service.adjudicate(
        normalized_claim,
        validation,
        matched_policies,
        provider_context=provider_context,
        utilization_context=utilization_context,
        payer_verification_context=payer_verification_context,
    )
    confidence_score = confidence_service.score(validation, decision)
    requires_human_review = (
        confidence_score < settings.human_review_threshold or decision.outcome == "review"
    )

    return repository.create_processed_claim(
        claim=normalized_claim,
        validation=validation,
        decision=decision,
        confidence_score=confidence_score,
        requires_human_review=requires_human_review,
        matched_policies=matched_policies,
        payer_verification_context=payer_verification_context,
    )


@router.get("/demo", response_model=ClaimSubmission)
def get_demo_claim() -> ClaimSubmission:
    return get_demo_outpatient_claim()


@router.get("", response_model=list[ClaimRecordSummary])
def list_claims(
    limit: int = Query(default=20, ge=1, le=100),
    offset: int = Query(default=0, ge=0),
    outcome: Optional[str] = Query(default=None),
    requires_review: Optional[bool] = Query(default=None),
    repository: ClaimsRepository = Depends(get_claims_repository),
) -> list[ClaimRecordSummary]:
    return repository.list_claims(
        limit=limit,
        offset=offset,
        outcome=outcome,
        requires_review=requires_review,
    )


@router.get("/{claim_id}", response_model=ClaimProcessingResponse)
def get_claim(
    claim_id: str,
    repository: ClaimsRepository = Depends(get_claims_repository),
) -> ClaimProcessingResponse:
    claim = repository.get_claim(claim_id)
    if claim is None:
        raise HTTPException(status_code=404, detail="Claim not found.")
    return claim


@router.post("/process", response_model=ClaimProcessingResponse)
def process_claim(
    claim: ClaimSubmission,
    settings: Settings = Depends(get_settings),
    repository: ClaimsRepository = Depends(get_claims_repository),
) -> ClaimProcessingResponse:
    return _run_claim_processing(claim, settings, repository)


@router.post("/intake-document", response_model=ClaimDocumentIntakeResponse)
async def intake_claim_document(
    file: UploadFile = File(...),
    auto_process: bool = Form(False),
    payer_name_hint: Optional[str] = Form(default=None),
    settings: Settings = Depends(get_settings),
    repository: ClaimsRepository = Depends(get_claims_repository),
) -> ClaimDocumentIntakeResponse:
    content = await file.read()
    try:
        intake_result = document_intake_service.extract_claim_draft(
            filename=file.filename or "claim-document",
            content=content,
            content_type=file.content_type,
            payer_name_hint=payer_name_hint,
        )
    except ClaimDocumentIntakeError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc

    if not auto_process or not intake_result.ready_for_processing:
        return intake_result

    claim, missing_fields = document_intake_service.finalize_claim_submission(intake_result.claim_draft)
    if claim is None:
        return intake_result.model_copy(
            update={
                "ready_for_processing": False,
                "missing_fields": missing_fields,
            }
        )

    processed_result = _run_claim_processing(claim, settings, repository)
    return intake_result.model_copy(
        update={
            "status": "processed",
            "processed_result": processed_result,
            "ready_for_processing": True,
            "missing_fields": [],
        }
    )


@router.post("/upload-x12", response_model=ClaimProcessingResponse)
async def upload_x12_claim(
    file: UploadFile = File(...),
    settings: Settings = Depends(get_settings),
    repository: ClaimsRepository = Depends(get_claims_repository),
) -> ClaimProcessingResponse:
    payload = (await file.read()).decode("utf-8", errors="ignore")
    try:
        claim = x12_parser.parse(payload)
    except X12ParseError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc

    return _run_claim_processing(claim, settings, repository)


@router.post("/upload-x12-batch", response_model=X12BatchUploadResponse)
async def upload_x12_batch(
    file: UploadFile = File(...),
    settings: Settings = Depends(get_settings),
    repository: ClaimsRepository = Depends(get_claims_repository),
) -> X12BatchUploadResponse:
    payload = (await file.read()).decode("utf-8", errors="ignore")
    try:
        claims = x12_parser.parse_many(payload)
    except X12ParseError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc

    results: list[X12BatchClaimResult] = []
    processed_claims = 0
    failed_claims = 0
    for claim in claims:
        try:
            processed = _run_claim_processing(claim, settings, repository)
        except Exception as exc:  # pragma: no cover - defensive batch isolation
            cleanup = getattr(repository, "cleanup_claim_by_external_id", None)
            if callable(cleanup):
                cleanup(claim.claim_id)
            failed_claims += 1
            results.append(
                X12BatchClaimResult(
                    claim_id=claim.claim_id,
                    status="failed",
                    error=str(exc),
                )
            )
            continue

        processed_claims += 1
        results.append(
            X12BatchClaimResult(
                claim_id=claim.claim_id,
                status="processed",
                result=processed,
            )
        )

    return X12BatchUploadResponse(
        total_claims=len(claims),
        processed_claims=processed_claims,
        failed_claims=failed_claims,
        results=results,
    )


@router.post("/{claim_id}/review", response_model=ClaimProcessingResponse)
def submit_claim_review(
    claim_id: str,
    review_request: ClaimReviewRequest,
    repository: ClaimsRepository = Depends(get_claims_repository),
) -> ClaimProcessingResponse:
    claim = repository.submit_review(claim_id, review_request)
    if claim is None:
        raise HTTPException(status_code=404, detail="Claim not found.")
    return claim

File: apps/web/src/shared/api/claims.ts
export type ServiceLine = {
  line_number: number;
  procedure_code: string;
  modifiers: string[];
  units: number;
  charge_amount: number;
};

export type ClaimSubmission = {
  claim_id: string;
  claim_type: "professional_outpatient";
  form_type: "CMS-1500";
  payer_name: string;
  plan_name: string;
  member_id: string;
  member_name: string;
  patient_id: string;
  provider_id: string;
  provider_name: string;
  place_of_service: string;
  diagnosis_codes: string[];
  procedure_codes: string[];
  service_lines: ServiceLine[];
  amount: number;
  date_of_service: string;
};

export type AuditEvent = {
  event_type: string;
  payload: Record<string, unknown>;
  created_at: string | null;
};

export type HumanReviewState = {
  status: string;
  reason: string;
  reviewer_name: string | null;
  reviewer_notes: string | null;
  updated_at: string | null;
};

export type InsightCard = {
  label: string;
  value: string;
  status: string;
  score: number;
  detail: string | null;
};

export type ClaimInsights = {
  policy_match: InsightCard;
  duplication_risk: InsightCard;
  network_parity: InsightCard;
};

export type ProviderAdjudicationContext = {
  provider_key: string;
  provider_name: string;
  specialty: string | null;
  network_status: string;
  contract_tier: string | null;
  contract_status: string;
  network_effective_date: string | null;
  network_end_date: string | null;
  participates_in_plan: boolean;
  specialty_match: boolean | null;
  specialty_match_reason: string | null;
};

export type UtilizationContext = {
  utilization_level: "routine" | "elevated" | "prior_auth";
  prior_auth_required: boolean;
  prior_auth_status: "not_required" | "pending_review" | "satisfied";
  trigger_codes: string[];
  review_reason: string | null;
  notes: string[];
};

export type ClaimReviewRequest = {
  reviewer_name?: string;
  reviewer_notes: string;
  review_status: "in_review" | "resolved";
  override_outcome?: "approve" | "deny" | "review";
};

export type AdjudicationCheck = {
  code: string;
  label: string;
  status: "passed" | "failed" | "review";
  source: "validation" | "policy" | "integrity" | "provider" | "utilization";
  summary: string;
};

export type ClaimDetailResponse = {
  claim: ClaimSubmission;
  status: string;
  created_at: string | null;
  confidence_score: number;
  requires_human_review: boolean;
  validation: {
    is_valid: boolean;
    issues: Array<{
      code: string;
      message: string;
      severity: string;
    }>;
  };
  decision: {
    outcome: string;
    rationale: string;
    cited_rules: string[];
    passed_checks: AdjudicationCheck[];
    failed_checks: AdjudicationCheck[];
    review_triggers: string[];
  };
  matched_policies: Array<{
    policy_id: string;
    title: string;
    summary: string;
    relevance_score: number;
    document_label?: string | null;
    source_reference?: string | null;
  }>;
  review_state: HumanReviewState | null;
  audit_trail: AuditEvent[];
  insights: ClaimInsights;
  provider_context: ProviderAdjudicationContext | null;
  utilization_context: UtilizationContext | null;
};

export type ClaimRecordSummary = {
  claim_id: string;
  claim_type: string;
  payer_name: string;
  member_name: string;
  provider_name: string;
  amount: number;
  date_of_service: string;
  outcome: string;
  confidence_score: number;
  requires_human_review: boolean;
  created_at: string;
  review_status: string | null;
};

export type X12BatchClaimResult = {
  claim_id: string | null;
  status: "processed" | "failed";
  result: ClaimDetailResponse | null;
  error: string | null;
};

export type X12BatchUploadResponse = {
  total_claims: number;
  processed_claims: number;
  failed_claims: number;
  results: X12BatchClaimResult[];
};

export type DraftServiceLine = {
  line_number: number | null;
  procedure_code: string | null;
  modifiers: string[];
  units: number | null;
  charge_amount: number | null;
};

export type ClaimDocumentDraft = {
  claim_id: string | null;
  claim_type: "professional_outpatient";
  form_type: "CMS-1500";
  payer_name: string | null;
  plan_name: string | null;
  member_id: string | null;
  member_name: string | null;
  patient_id: string | null;
  provider_id: string | null;
  provider_name: string | null;
  place_of_service: string | null;
  diagnosis_codes: string[];
  procedure_codes: string[];
  service_lines: DraftServiceLine[];
  amount: number | null;
  date_of_service: string | null;
};

export type LowConfidenceField = {
  field: string;
  confidence: "low" | "medium" | "high";
  reason: string;
};

export type ClaimDocumentIntakeResponse = {
  status: "drafted" | "processed";
  source_type: "text" | "pdf" | "docx" | "image";
  extraction_summary: string;
  claim_draft: ClaimDocumentDraft;
  ready_for_processing: boolean;
  missing_fields: string[];
  review_notes: string[];
  low_confidence_fields: LowConfidenceField[];
  processed_result: ClaimDetailResponse | null;
};

export type ClaimsFilter = {
  limit?: number;
  offset?: number;
  outcome?: string;
  requires_review?: boolean;
};

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL ?? "http://localhost:8000/api";

export async function fetchDemoClaim(): Promise<ClaimSubmission> {
  const response = await fetch(`${API_BASE_URL}/claims/demo`);
  if (!response.ok) throw new Error("Unable to fetch demo claim.");
  return response.json();
}

export async function fetchClaims(filter: ClaimsFilter = {}): Promise<ClaimRecordSummary[]> {
  const params = new URLSearchParams();
  if (filter.limit !== undefined) params.set("limit", String(filter.limit));
  if (filter.offset !== undefined) params.set("offset", String(filter.offset));
  if (filter.outcome !== undefined) params.set("outcome", filter.outcome);
  if (filter.requires_review !== undefined)
    params.set("requires_review", String(filter.requires_review));

  const qs = params.toString();
  const response = await fetch(`${API_BASE_URL}/claims${qs ? `?${qs}` : ""}`);
  if (!response.ok) throw new Error("Unable to fetch claims.");
  return response.json();
}

export async function fetchClaimById(claimId: string): Promise<ClaimDetailResponse> {
  const response = await fetch(`${API_BASE_URL}/claims/${claimId}`);
  if (!response.ok) throw new Error("Unable to fetch claim detail.");
  return response.json();
}

export async function processClaim(claim: ClaimSubmission): Promise<ClaimDetailResponse> {
  const response = await fetch(`${API_BASE_URL}/claims/process`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(claim),
  });
  if (!response.ok) throw new Error("Unable to process claim.");
  return response.json();
}

export async function submitReview(
  claimId: string,
  review: ClaimReviewRequest,
): Promise<ClaimDetailResponse> {
  const response = await fetch(`${API_BASE_URL}/claims/${claimId}/review`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(review),
  });
  if (!response.ok) throw new Error("Unable to submit review.");
  return response.json();
}

export async function uploadX12Claim(file: File): Promise<ClaimDetailResponse> {
  const form = new FormData();
  form.append("file", file);
  const response = await fetch(`${API_BASE_URL}/claims/upload-x12`, {
    method: "POST",
    body: form,
  });
  if (!response.ok) {
    const data: unknown = await response.json().catch(() => ({}));
    const detail = (data as { detail?: string }).detail;
    throw new Error(detail ?? "Unable to process X12 file.");
  }
  return response.json();
}

export async function uploadX12Batch(file: File): Promise<X12BatchUploadResponse> {
  const form = new FormData();
  form.append("file", file);
  const response = await fetch(`${API_BASE_URL}/claims/upload-x12-batch`, {
    method: "POST",
    body: form,
  });
  if (!response.ok) {
    const data: unknown = await response.json().catch(() => ({}));
    const detail = (data as { detail?: string }).detail;
    throw new Error(detail ?? "Unable to process X12 batch file.");
  }
  return response.json();
}

export async function intakeClaimDocument(
  file: File,
  options: { autoProcess?: boolean; payerNameHint?: string } = {},
): Promise<ClaimDocumentIntakeResponse> {
  const form = new FormData();
  form.append("file", file);
  form.append("auto_process", String(options.autoProcess ?? false));
  if (options.payerNameHint) {
    form.append("payer_name_hint", options.payerNameHint);
  }
  const response = await fetch(`${API_BASE_URL}/claims/intake-document`, {
    method: "POST",
    body: form,
  });
  if (!response.ok) {
    const data: unknown = await response.json().catch(() => ({}));
    const detail = (data as { detail?: string }).detail;
    throw new Error(detail ?? "Unable to extract claim document.");
  }
  return response.json();
}
